File: parking_app/lib/models/booking.dart
import 'package:parking_app/models/parking.dart';
import 'package:parking_app/models/vehicle.dart';

class Booking {
  final int id;
  final int userId;
  final int parkingLocationId;
  final int vehicleId;
  final DateTime startTime;
  final DateTime endTime;
  final String status;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int durationHours;
  final double totalPrice;
  final Parking parkingLocation;
  final Vehicle vehicle;

  Booking({
    required this.id,
    required this.userId,
    required this.parkingLocationId,
    required this.vehicleId,
    required this.startTime,
    required this.endTime,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    required this.durationHours,
    required this.totalPrice,
    required this.parkingLocation,
    required this.vehicle,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'] as int,
      userId: json['user_id'] as int,
      parkingLocationId: json['parking_location_id'] as int,
      vehicleId: json['vehicle_id'] as int,
      startTime: DateTime.parse(json['start_time'] as String),
      endTime: DateTime.parse(json['end_time'] as String),
      status: json['status'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      durationHours: json['duration_hours'] as int,
      totalPrice: double.parse(json['total_price'] as String),
      parkingLocation: Parking.fromJson(
        json['parking_location'] as Map<String, dynamic>,
      ),
      vehicle: Vehicle.fromJson(json['vehicle'] as Map<String, dynamic>),
    );
  }
}

File: parking-app-backend/app/Models/Booking.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Booking extends Model
{
    use HasFactory;
    

    protected $fillable = [
        'user_id',
        'parking_location_id',
        'vehicle_id',
        'start_time',
        'end_time',
        'status',
        'duration_hours',
        'total_price'
    ];
    

    protected $casts = [
        'start_time' => 'datetime',
        'end_time' => 'datetime',
        'duration_hours' => 'integer',
        'total_price' => 'decimal:2',
    ];
    
    /**
     * The status of the booking.
     */
    const STATUS_BOOKED = 'booked';
    const STATUS_IN_PROGRESS = 'in_progress';
    const STATUS_COMPLETED = 'completed';
    const STATUS_CANCELLED = 'cancelled';
    

    public function user()
    {
        return $this->belongsTo(User::class);
    }
    
    /**
     * Get the parking location for this booking.
     */
    public function parkingLocation()
    {
        return $this->belongsTo(ParkingLocation::class);
    }
    
    /**
     * Get the vehicle for this booking.
     */
    public function vehicle()
    {
        return $this->belongsTo(Vehicle::class);
    }
    
  
    public function scopeActive($query)
    {
        return $query->whereIn('status', [self::STATUS_BOOKED, self::STATUS_IN_PROGRESS]);
    }
    

    public function isActive()
    {
        return in_array($this->status, [self::STATUS_BOOKED, self::STATUS_IN_PROGRESS]);
    }
    

    public function canBeExtended()
    {
        return $this->isActive() && !now()->isAfter($this->end_time);
    }
    
    public function canBeCancelled()
    {
        return $this->status === self::STATUS_BOOKED && now()->isBefore($this->start_time);
    }
}
